net/fetcher.go

```diff
@@ -22,6 +30,212 @@ type Fetcher interface {
+	gsf := &GraphSyncFetcher{
+		bs:        blockstore,
+		validator: bv,
+		bridge:    bridge,
+	}
+	gsf.gs = graphsync.New(ctx, network, bridge, gsf.loader, gsf.storer)
+	return gsf
+}
+
+func (gsf *GraphSyncFetcher) loader(lnk ipldp.Link, lnkCtx ipldp.LinkContext) (io.Reader, error) {
+	asCidLink, ok := lnk.(cidlink.Link)
+	if !ok {
+		return nil, fmt.Errorf("Unsupported Link Type")
+	}
+	block, err := gsf.bs.Get(asCidLink.Cid)
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewReader(block.RawData()), nil
+}
+
+func (gsf *GraphSyncFetcher) storer(lnkCtx ipldp.LinkContext) (io.Writer, ipldp.StoreCommitter, error) {
+	var buffer bytes.Buffer
+	committer := func(lnk ipldp.Link) error {
+		asCidLink, ok := lnk.(cidlink.Link)
+		if !ok {
+			return fmt.Errorf("Unsupported Link Type")
+		}
+		block, err := blocks.NewBlockWithCid(buffer.Bytes(), asCidLink.Cid)
+		if err != nil {
+			return err
+		}
+		return gsf.bs.Put(block)
+	}
+	return &buffer, committer, nil
+}
+
+const maxRecursionDepth = 64
+const recursionMultiplier = 4
+
+// FetchTipSets gets Tipsets starting from the given tipset key and continuing till
+// the done function returns true or errors
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	err := gsf.GetSingleLayer(ctx, cids, from)
+	if err != nil {
+		return nil, err
+	}
+	startingTipset, err := gsf.LoadTipsetFromCids(ctx, cids)
+	if err != nil {
+		return nil, err
+	}
+	out := []types.TipSet{startingTipset}
+	isDone, err := done(startingTipset)
+	if err != nil {
+		return nil, err
+	}
+	if !isDone {
+		baseCid := cids[0]
+		recursionDepth := 1
+	makeRecursion:
+		for {
+			cidResponseLayers, err := gsf.GetBlocks(ctx, baseCid, from, recursionDepth)
+			if err != nil {
+				return nil, err
+			}
+			for _, cidResponseLayer := range cidResponseLayers {
+				ts, err := gsf.LoadTipsetFromCids(ctx, cidResponseLayer)
+				if err != nil {
+					return nil, err
+				}
+
+				out = append(out, ts)
+				isDone, err := done(ts)
+				if err != nil {
+					return nil, err
+				}
+
+				if isDone {
+					break makeRecursion
+				}
+				baseCid = cidResponseLayer[0]
+			}
+			if recursionDepth < maxRecursionDepth {
+				recursionDepth *= recursionMultiplier
+			}
+		}
+	}
+	return out, nil
+}
+
+// LoadTipsetFromCids loads and sanitizes an already downloaded TipSet
+// from an array of cids
+func (gsf *GraphSyncFetcher) LoadTipsetFromCids(ctx context.Context, cidResponseLayer []cid.Cid) (types.TipSet, error) {
+	blockResponseLayer := make([]blocks.Block, 0, len(cidResponseLayer))
+	for _, cid := range cidResponseLayer {
+		block, err := gsf.bs.Get(cid)
+		if err != nil {
+			return types.UndefTipSet, err
+		}
+		blockResponseLayer = append(blockResponseLayer, block)
+	}
+	validatedBlocks, err := sanitizeBlocks(ctx, blockResponseLayer, gsf.validator)
```

Comment:
This implies that blocks are written to the store without being validated, and that seems bad to me. Can we validate in the storer instead? Or is the intention to avoid deserialising the stored bytes?

Reply:
yea I was trying to minimize the number serializations/deserializations to store. 

One thing I would point out is that while it may not be obvious -- Bitswap is doing the same thing:
https://github.com/ipfs/go-bitswap/blob/master/bitswap.go#L282
That's in the code path every time you get a wanted block from the network.

Arguably with Bitswap it's less bad because only one layer at a time. 

One other complication of putting things in the storer is that currently the loader/storer can still be extracted pretty easily from the fetcher so that Graphsync is initialized outside the fetcher. Which, again, was the next step. However, this is more complicated if we're trying to do validation inside the storer. Though I guess if we're don't mind the extra serialize/deserialize we can just return an error in the storer when we're unable to verify format which should abort the Graphsync call with an error.

Reply:
Guarding block store stores with a custom validation function sounds great for the future, but as @hannahhoward says it would be a new feature.  Unless it is very simple we should wait for another PR to make this improvement.

Reply:
Also, longer term I believe we may be able to use some neat facilities of `go-ipld-prime` to do this without extra serialization/deserialization, when we are ready to start integrating parts of it into the rest of the codebase.

Reply:
Happy for your discretion, but please leave a comment describing this to the future.

Aside: my preference for in-code documentation is definitely for more than the average.